signer/src/block_notifier.rs

```diff
@@ -0,0 +1,150 @@
+//! This module provides functionality for notifying about new blocks using the Electrum client.
+
+use crate::config::SETTINGS;
+use electrum_client::bitcoin::BlockHash;
+use electrum_client::{
+    Client, ConfigBuilder, ElectrumApi, Error as ElectrumError, HeaderNotification,
+};
+use futures::stream::Stream;
+use std::error::Error;
+use std::pin::Pin;
+use std::sync::Arc;
+use tokio::sync::mpsc::{channel, Sender};
+use tokio::task;
+use tokio::time::{interval, sleep, Duration};
+use tokio_stream::wrappers::ReceiverStream;
+use tracing::{error, warn};
+
+/// The `BlockNotifier` trait defines a method for subscribing to a stream of block headers.
+pub trait BlockNotifier {
+    /// Errors occurring during subscription or running the notifier.
+    type Error: Error;
+
+    /// Returns a stream of block headers.
+    fn subscribe(&self) -> Pin<Box<dyn Stream<Item = Result<BlockHash, Self::Error>> + Send>>;
```

Comment:
While this is alright, I'd prefer to use generics over trait objects so we don't have to do explicit boxing and pinning
```suggestion
    fn subscribe(&self) -> impl Stream<Item = Result<BlockHash, Self::Error>> + Send;
```
this makes the code a bit cleaner (in my opinion) and saves us some redundant heap allocations.